### ISSUE TITLE
Underlying implementation for explore_adf softmax

### ISSUE THREAD

**Issue by paulusm** [VowpalWabbit/vowpal_wabbit#4711]
### Description

I've gone over the docs a few times, but I'm unclear as to which bandit evaluation algorithm underlies the explore_adf softmax explorer, i.e. how is the quality of an action predicted?

### Link to Documentation Page

https://github.com/VowpalWabbit/vowpal_wabbit/wiki/Contextual-Bandit-algorithms

**Comment by sm-28601** [comment]
The explore_adf softmax reduction applies softmax exploration over cost-based predictions from a base learner. Lower-cost actions receive higher probabilities. A lambda parameter sharpens the distribution, and a minimum epsilon probability is enforced per action.

**Comment by paulusm** [comment]
Thanks Sahil. My question is : What is the form of the base learner model?

**Comment by JohnLangford** [comment]
## Softmax Explorer Algorithm Explanation

The `--softmax` explorer in `cb_explore_adf` works as follows:

### Underlying Action Quality Prediction

The softmax explorer uses the base **cb_adf** (Contextual Bandit with Action-Dependent Features) learner, which predicts a **cost score** for each action using online regression. The cost represents how "bad" an action is expected to be - lower cost = better action.

For each context-action pair, cb_adf learns:
- Feature weights through stochastic gradient descent
- The predicted cost is the dot product of features and learned weights

### Converting Costs to Probabilities

The softmax transformation converts predicted costs to selection probabilities:

```
P(action_i) ∝ exp(-λ * cost(action_i))
```

Where:
- **λ (lambda)**: Temperature parameter (default 1.0, set via `--lambda`)
  - Higher λ → more greedy (concentrate probability on lowest-cost action)
  - Lower λ → more uniform exploration
  - λ = 0 → uniform random selection

The implementation normalizes these values to create a valid probability distribution.

### Minimum Probability Enforcement

After the softmax transformation, the `--epsilon` parameter (if specified) ensures each action has at least `epsilon / num_actions` probability, guaranteeing exploration.

### Summary

1. **cb_adf** learns cost predictions for each action using features
2. **Softmax** transforms costs → probabilities (lower cost = higher prob)
3. **Epsilon** (optional) enforces minimum exploration

### Code References

- Softmax explorer: `vowpalwabbit/core/src/reductions/cb/cb_explore_adf_softmax.cc`
- Softmax formula: `vowpalwabbit/explore/include/vw/explore/explore_internal.h` (lines 47-87)
- Base cb_adf learner: `vowpalwabbit/core/src/reductions/cb/cb_adf.cc`

I'd recommend updating the [Contextual Bandit algorithms wiki page](https://github.com/VowpalWabbit/vowpal_wabbit/wiki/Contextual-Bandit-algorithms) to include this explanation.

**Comment by JohnLangford** [comment]
The wiki documentation has been updated with a detailed explanation of how the softmax explorer works:

**Updated page:** [Contextual Bandit algorithms](https://github.com/VowpalWabbit/vowpal_wabbit/wiki/Contextual-Bandit-algorithms)

The new documentation covers:
- How action quality is predicted (cost-sensitive regression via online SGD)
- The softmax formula: P(a) = exp(-λ × score) / Σ exp(-λ × score_i)
- Lambda parameter effects (λ=0 for uniform, λ→∞ for greedy)
- How epsilon provides minimum exploration guarantee

Feel free to reopen if you have additional questions.
